packages/base/src/api/client.ts

```diff
@@ -19,17 +22,35 @@ export abstract class BaseApiClient {
     this.apiKey = params.apiKey;
     this.apiSecret = params.apiSecret;
     this.httpsAgent = params.httpsAgent;
+    this.session = undefined;
   }
 
   protected async init(): Promise<AxiosInstance> {
     if (!this.api) {
       const userPass = { Username: this.apiKey, Password: this.apiSecret };
       const poolData = { UserPoolId: this.getPoolId(), ClientId: this.getPoolClientId() };
-      this.api = createAuthenticatedApi(userPass, poolData, this.getApiUrl(), this.httpsAgent);
+      this.session = await authenticate(userPass, poolData);
+      this.api = createAuthenticatedApi(userPass.Username, this.session, this.getApiUrl(), this.httpsAgent);
     }
     return this.api;
   }
 
+  protected async refresh(): Promise<AxiosInstance> {
+    if (!this.session) {
+      return this.init();
+    }
+    try {
+      const userPass = { Username: this.apiKey, Password: this.apiSecret };
+      const poolData = { UserPoolId: this.getPoolId(), ClientId: this.getPoolClientId() };
+      this.session = await refreshSession(userPass, poolData, this.session);
```

Comment:
super minor optimization as this will only happen very rarely for super long-running sessions (and not worth it if it's a big task), but I think we could calculate the expiration on the refresh token and call `this.init()` if we are already past expiration rather than relying on the `catch` alone.